cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,331 @@
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    Long lastModifiedTimestamp = getCurrentLastModifiedTimestamp(artifactId);
+
+    File newJarLocation = Retries
+      .callWithRetries(() -> fetchArtifact(artifactId, lastModifiedTimestamp), retryStrategy);
+
+    // If the lastModifiedTimestamp is not null then we might need to delete the old cache.
+    if (lastModifiedTimestamp != null) {
+      // TODO (CDAP-18051): Migrate this cleanup task to its own service
+      deleteOldCache(artifactId, lastModifiedTimestamp, newJarLocation);
+    }
+    return newJarLocation;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesn't exist, creating it now");
+      BundleJarUtil.unJar(jarLocation, unpackDir);
+    }
+    return unpackDir;
+  }
+
+  /**
+   * fetchArtifact attempts to connect to app fabric to download the given artifact. This method will throw {@link
+   * RetryableException} in certain circumstances so using this with the
+   */
+  private File fetchArtifact(ArtifactId artifactId,
+                             @Nullable Long lastModifiedTimestamp) throws IOException, ArtifactNotFoundException {
+
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = NamespaceId.DEFAULT.getEntityName();
+      scope = ArtifactScope.SYSTEM;
+    }
+
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getArtifact(),
+                               artifactId.getVersion(),
+                               scope);
+
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
+    if (lastModifiedTimestamp != null) {
```

Comment:
The `try` should start right after the `openConnection` call.

Reply:
done